### PR TITLE
test(ctb): Add failure tests for finalizeWithdrawalTransaction

### PR DESCRIPTION
**Pull request by maurelian**
Adds a test for each revert occurring in `finalizeWithdrawalTransaction`. 

Fixes ENG-2474

### CODE REVIEW

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 1]
```diff
@@ -215,62 +214,14 @@ contract OptimismPortal_Test is Portal_Initializer {
...
-
-        vm.warp(
-            oracle.getL2Output(oracle.latestBlockNumber()).timestamp +
-                op.FINALIZATION_PERIOD_SECONDS() + 1
-        );
-
-        vm.expectRevert("OptimismPortal: invalid output root proof");
-        op.finalizeWithdrawalTransaction(Types.WithdrawalTransaction(0, alice, alice, 0, 0, hex""), 0, outputRootProof, hex"");
-    }
+    // TODO: test this deeply
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:217]: Why is this TODO being added back in?

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 2]
```diff
@@ -318,43 +505,33 @@ contract OptimismPortal_Test is Portal_Initializer {
         uint8 _gasLimit,
         bytes memory _data
     ) external {
-        // Cannot call the optimism portal
-        vm.assume(_target != address(op));
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:322]: Why did you remove this `vm.assume`? It will eventually find an invalid input without it

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 3]
```diff
@@ -318,43 +505,33 @@ contract OptimismPortal_Test is Portal_Initializer {
...
         uint256 _nonce = messagePasser.nonce();
-
+        Types.WithdrawalTransaction memory _tx = Types.WithdrawalTransaction({
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:509]: This test is annoying because of stack too deep errors, it looks like using the struct here helps

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 4]
```diff
@@ -310,6 +261,242 @@ contract OptimismPortal_Test is Portal_Initializer {
...
+        uint256 finalizedTimestamp = block.timestamp - op.FINALIZATION_PERIOD_SECONDS() - 1;
+        bytes memory data = abi.encodeWithSelector(this.callAndExpectRevert.selector);
+        (
+            bytes32 stateRoot,
+            bytes32 storageRoot,
+            bytes32 outputRoot,
+            bytes32 withdrawalHash,
+            bytes memory withdrawalProof
+        ) = ffi.getFinalizeWithdrawalTransactionInputs(_defaultTx);
+        // ) = getFinalizeWithdrawalTransactionInputs(_n, _s, address(this), 0, _g, data);
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:480]: Dead code?

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 5]
```diff
@@ -310,6 +261,242 @@ contract OptimismPortal_Test is Portal_Initializer {
...
+            stateRoot: stateRoot,
+            withdrawerStorageRoot: storageRoot,
+            latestBlockhash: bytes32(0)
+        });
+        vm.mockCall(
+            address(op.L2_ORACLE()),
+            abi.encodeWithSelector(L2OutputOracle.getL2Output.selector),
+            abi.encode(Types.OutputProposal(outputRoot, finalizedTimestamp))
+        );
+
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:492]: Could you add a `vm.expectCall` here? Its not clear what the assertion is that is being made just by reading the test

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 6]
```diff
@@ -310,6 +261,242 @@ contract OptimismPortal_Test is Portal_Initializer {
...
+        assert(address(bob).balance == bobBalanceBefore + 100);
+    }
+
+    // Test: finalizeWithdrawalTransaction cannot finalize a withdrawal with itself (the OptimismPortal) as the target.
+    function test_finalizeWithdrawalTransaction_revertsOnSelfCall() external {
+        _defaultTx.target = address(op);
+        vm.expectRevert("OptimismPortal: you cannot send messages to the portal contract");
+        op.finalizeWithdrawalTransaction(
+            _defaultTx,
+            // _n,
```
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:342]: Is this dead code?

**Review by tynes — CHANGES_REQUESTED** [review]
Left a bunch of comments

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol (hunk 3 above)
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:534]: Makes the code cleaner IMO too.

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol (hunk 2 above)
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:322]: My mistake.

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol (hunk 5 above)
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:508]: Yeah, this test is pretty hard to follow (and was actually broken before). 
I've added several checks with `vm.expectX()` and `assert()`, as well as comments to explain what's going on.

**Review by tynes — APPROVED** [review]

File: packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol [hunk 7]
```diff
@@ -215,62 +214,11 @@ contract OptimismPortal_Test is Portal_Initializer {
         assertEq(address(op).balance, NON_ZERO_VALUE);
     }
 
-    function test_cannotVerifyRecentWithdrawal() external {
```
Comment (maurelian) [inline packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol:218]: Not removed, this was just relocated into the new test contract and renamed to `test_finalizeWithdrawalTransaction_revertsOnRecentWithdrawal()`.

**Review by protolambda — APPROVED** [review]

**Comment by tynes** [comment]
Looks like CI found a counterexample:

```
[FAIL. Counterexample: calldata=0x9c7e352a000000000000000000000000b4c79dab8f259c7aee6e5b2aa729821864227e84000000000000000000000000028188030556faf8383a2ba80d95e1415c6ed6f6000000000000000000000000000000000000000000000000541b834cd1f08f0700000000000000000000000000000000000000000000000000000000000000a400000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000020945a53f5ffffffffffffffff0000000000000000000000000000000000000000, args=[0xb4c79dab8f259c7aee6e5b2aa729821864227e84, 0x028188030556faf8383a2ba80d95e1415c6ed6f6, 6060582089520090887, 164, 0x945a53f5ffffffffffffffff0000000000000000000000000000000000000000]] test_finalizeWithdrawalTransaction_differential(address,address,uint64,uint8,bytes) (runs: 225, μ: 252441, ~: 266586)
```
